File: app/src/main/java/deltazero/amarok/ui/CountdownConfirmDialog.java
package deltazero.amarok.ui;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;

import androidx.annotation.NonNull;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;

import deltazero.amarok.R;

public class CountdownConfirmDialog {

    private final Context context;
    private final String title;
    private final String message;
    private final int countdownSeconds;
    private final Runnable onConfirmAction;
    private final Runnable onCancelAction;

    private CountdownConfirmDialog(Builder builder) {
        this.context = builder.context;
        this.title = builder.title;
        this.message = builder.message;
        this.countdownSeconds = builder.countdownSeconds;
        this.onConfirmAction = builder.onConfirmAction;
        this.onCancelAction = builder.onCancelAction;
    }

    public void show() {
        MaterialAlertDialogBuilder dialogBuilder = new MaterialAlertDialogBuilder(context);
        dialogBuilder.setTitle(title);
        dialogBuilder.setMessage(message);

        // Positive button (initially disabled)
        dialogBuilder.setPositiveButton(android.R.string.ok, (d, which) -> {
            if (onConfirmAction != null) {
                onConfirmAction.run();
            }
        });

        // Negative button
        dialogBuilder.setNegativeButton(android.R.string.cancel, (d, which) -> {
            if (onCancelAction != null) {
                onCancelAction.run();
            }
        });

        var dialog = dialogBuilder.create();
        dialog.show();

        // Disable the positive button initially and set countdown text
        var positiveButton = dialog.getButton(androidx.appcompat.app.AlertDialog.BUTTON_POSITIVE);
        if (positiveButton != null) {
            positiveButton.setEnabled(false);
            positiveButton.setText(context.getString(R.string.confirm_with_countdown, countdownSeconds));
        }

        // Start countdown
        final var finalDialog = dialog;
        Handler handler = new Handler(Looper.getMainLooper());
        final int[] countdown = {countdownSeconds};

        Runnable countdownRunnable = new Runnable() {
            @Override
            public void run() {
                if (countdown[0] > 0 && finalDialog.isShowing()) {
                    countdown[0]--;
                    var button = finalDialog.getButton(androidx.appcompat.app.AlertDialog.BUTTON_POSITIVE);
                    if (button != null) {
                        if (countdown[0] > 0) {
                            button.setText(context.getString(R.string.confirm_with_countdown, countdown[0]));
                        } else {
                            button.setText(context.getString(R.string.confirm));
                            button.setEnabled(true);
                        }
                    }
                    if (countdown[0] > 0) {
                        handler.postDelayed(this, 1000);
                    }
                }
            }
        };

        handler.postDelayed(countdownRunnable, 1000);
    }

    public static class Builder {
        private final Context context;
        private String title = "";
        private String message = "";
        private int countdownSeconds = 5;
        private Runnable onConfirmAction;
        private Runnable onCancelAction;

        public Builder(@NonNull Context context) {
            this.context = context;
        }

        public Builder setTitle(String title) {
            this.title = title;
            return this;
        }

        public Builder setTitle(int titleResId) {
            this.title = context.getString(titleResId);
            return this;
        }

        public Builder setMessage(String message) {
            this.message = message;
            return this;
        }

        public Builder setMessage(int messageResId) {
            this.message = context.getString(messageResId);
            return this;
        }

        public Builder setCountdownSeconds(int seconds) {
            this.countdownSeconds = seconds;
            return this;
        }

        public Builder setOnConfirmAction(Runnable action) {
            this.onConfirmAction = action;
            return this;
        }

        public Builder setOnCancelAction(Runnable action) {
            this.onCancelAction = action;
            return this;
        }

        public CountdownConfirmDialog build() {
            return new CountdownConfirmDialog(this);
        }

        public void show() {
            build().show();
        }
    }
}

File: app/src/main/java/deltazero/amarok/ui/CountdownConfirmDialog.kt
package deltazero.amarok.ui

import android.content.Context
import android.os.Handler
import android.os.Looper
import androidx.appcompat.app.AlertDialog
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import deltazero.amarok.R

class CountdownConfirmDialog private constructor(builder: Builder) {
  private val context: Context = builder.context
  private val title: String = builder.title
  private val message: String = builder.message
  private val countdownSeconds: Int = builder.countdownSeconds
  private val onConfirmAction: Runnable? = builder.onConfirmAction
  private val onCancelAction: Runnable? = builder.onCancelAction

  fun show() {
    val dialogBuilder = MaterialAlertDialogBuilder(context)
    dialogBuilder.setTitle(title)
    dialogBuilder.setMessage(message)

    // Positive button (initially disabled)
    dialogBuilder.setPositiveButton(android.R.string.ok) { _, _ -> onConfirmAction?.run() }

    // Negative button
    dialogBuilder.setNegativeButton(android.R.string.cancel) { _, _ -> onCancelAction?.run() }

    val dialog = dialogBuilder.create()
    dialog.show()

    // Disable the positive button initially and set countdown text
    val positiveButton = dialog.getButton(AlertDialog.BUTTON_POSITIVE)
    if (positiveButton != null) {
      positiveButton.isEnabled = false
      positiveButton.text = context.getString(R.string.confirm_with_countdown, countdownSeconds)
    }

    // Start countdown
    val finalDialog = dialog
    val handler = Handler(Looper.getMainLooper())
    val countdown = intArrayOf(countdownSeconds)

    val countdownRunnable =
      object : Runnable {
        override fun run() {
          if (countdown[0] > 0 && finalDialog.isShowing) {
            countdown[0]--
            val button = finalDialog.getButton(AlertDialog.BUTTON_POSITIVE)
            if (button != null) {
              if (countdown[0] > 0) {
                button.text = context.getString(R.string.confirm_with_countdown, countdown[0])
              } else {
                button.text = context.getString(R.string.confirm)
                button.isEnabled = true
              }
            }
            if (countdown[0] > 0) {
              handler.postDelayed(this, 1000)
            }
          }
        }
      }

    handler.postDelayed(countdownRunnable, 1000)
  }

  class Builder internal constructor(internal val context: Context) {
    internal var title: String = ""
    internal var message: String = ""
    internal var countdownSeconds: Int = 5
    internal var onConfirmAction: Runnable? = null
    internal var onCancelAction: Runnable? = null

    fun setTitle(title: String): Builder {
      this.title = title
      return this
    }

    fun setTitle(titleResId: Int): Builder {
      title = context.getString(titleResId)
      return this
    }

    fun setMessage(message: String): Builder {
      this.message = message
      return this
    }

    fun setMessage(messageResId: Int): Builder {
      message = context.getString(messageResId)
      return this
    }

    fun setCountdownSeconds(seconds: Int): Builder {
      countdownSeconds = seconds
      return this
    }

    fun setOnConfirmAction(action: Runnable?): Builder {
      onConfirmAction = action
      return this
    }

    fun setOnCancelAction(action: Runnable?): Builder {
      onCancelAction = action
      return this
    }

    fun build(): CountdownConfirmDialog = CountdownConfirmDialog(this)

    fun show() {
      build().show()
    }
  }
}
